{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"eval_on_coronavirus.ipynb","provenance":[],"collapsed_sections":[],"machine_shape":"hm","authorship_tag":"ABX9TyOr1XoSHdZAI3dxrVqF5wRI"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"code","metadata":{"id":"_vp_He8fp65h","colab_type":"code","outputId":"5746e61a-20b3-44e5-dcbb-997d2888de2f","executionInfo":{"status":"ok","timestamp":1591771925987,"user_tz":420,"elapsed":17002,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":122}},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":1,"outputs":[{"output_type":"stream","text":["Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly\n","\n","Enter your authorization code:\n","··········\n","Mounted at /content/drive\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"9ma7foXyp8Xz","colab_type":"code","outputId":"d90bbcac-0719-4efb-9b90-1fb403151f53","executionInfo":{"status":"ok","timestamp":1591772028588,"user_tz":420,"elapsed":101349,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":323}},"source":["!pip3 install torch==1.2.0+cu92 torchvision==0.4.0+cu92 -f https://download.pytorch.org/whl/torch_stable.html"],"execution_count":2,"outputs":[{"output_type":"stream","text":["Looking in links: https://download.pytorch.org/whl/torch_stable.html\n","Collecting torch==1.2.0+cu92\n","\u001b[?25l  Downloading https://download.pytorch.org/whl/cu92/torch-1.2.0%2Bcu92-cp36-cp36m-manylinux1_x86_64.whl (663.1MB)\n","\u001b[K     |████████████████████████████████| 663.1MB 23kB/s \n","\u001b[?25hCollecting torchvision==0.4.0+cu92\n","\u001b[?25l  Downloading https://download.pytorch.org/whl/cu92/torchvision-0.4.0%2Bcu92-cp36-cp36m-manylinux1_x86_64.whl (8.8MB)\n","\u001b[K     |████████████████████████████████| 8.8MB 31.7MB/s \n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (from torch==1.2.0+cu92) (1.18.5)\n","Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from torchvision==0.4.0+cu92) (1.12.0)\n","Requirement already satisfied: pillow>=4.1.1 in /usr/local/lib/python3.6/dist-packages (from torchvision==0.4.0+cu92) (7.0.0)\n","Installing collected packages: torch, torchvision\n","  Found existing installation: torch 1.5.0+cu101\n","    Uninstalling torch-1.5.0+cu101:\n","      Successfully uninstalled torch-1.5.0+cu101\n","  Found existing installation: torchvision 0.6.0+cu101\n","    Uninstalling torchvision-0.6.0+cu101:\n","      Successfully uninstalled torchvision-0.6.0+cu101\n","Successfully installed torch-1.2.0+cu92 torchvision-0.4.0+cu92\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"-QEQQqD2fvEH","colab_type":"code","colab":{}},"source":["import os\n","os.chdir('/content/drive/My Drive/Peptide MHC Project')"],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"Ii_1ypTHp_BP","colab_type":"code","outputId":"8d1ffe10-4fbe-4d83-e8c2-4bf8657671f5","executionInfo":{"status":"ok","timestamp":1588888984082,"user_tz":420,"elapsed":861,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":34}},"source":["import torch\n","import torch.nn as nn \n","from architecture import *\n","import pandas as pd\n","import numpy as np\n","\n","print(torch.__version__)"],"execution_count":0,"outputs":[{"output_type":"stream","text":["1.2.0+cu92\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"ZN3Kd4eUp130","colab_type":"code","outputId":"51b486a8-7760-4043-8942-a496715b3c49","executionInfo":{"status":"ok","timestamp":1588888987767,"user_tz":420,"elapsed":965,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":51}},"source":["# Loads Coronavirus Peptide Sequences\n","# Full data (in all) has too many peptides of length 11-15\n","corona_embedding_tensor = torch.load('coronavirus/subset_corona_embeddings.pt').cuda()\n","print(corona_embedding_tensor.shape)\n","corona_peptides = pd.read_csv(\"coronavirus/subset_corona_peptides.txt\", header=None)\n","corona_peptides.columns = ['peptide']\n","\n","# Loads MHC Allele Sequences\n","alpha_140_179_embedding_tensor = torch.load('embeddings/common_140_179_embeddings.pt')\n","alpha_50_84_embedding_tensor = torch.load('embeddings/common_50_84_embeddings.pt')\n","allele_embedding_tensor = torch.cat((alpha_140_179_embedding_tensor, alpha_50_84_embedding_tensor), dim=1).cuda()\n","print(allele_embedding_tensor.shape)\n","alleles = pd.read_csv(\"allele_sequences/30_mhc_alleles.csv\")\n"],"execution_count":0,"outputs":[{"output_type":"stream","text":["torch.Size([625, 15, 100])\n","torch.Size([30, 106, 100])\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"rPfVM-nqp2zE","colab_type":"code","outputId":"96075ed6-01ad-42b6-c759-092539472e0d","executionInfo":{"status":"ok","timestamp":1588888989761,"user_tz":420,"elapsed":909,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":34}},"source":["# Defines model\n","peptide_embedding_dim = 100\n","peptide_hidden_dim = 100\n","\n","allele_embedding_dim = 100\n","allele_hidden_dim = 100\n","\n","peptide_to_output_dim = peptide_hidden_dim*2*2\n","allele_to_output_dim = allele_hidden_dim*2*2\n","\n","peptide_model = peptide_BiGRU(peptide_embedding_dim, peptide_hidden_dim).cuda()\n","allele_model = allele_BiGRU(allele_embedding_dim, allele_hidden_dim).cuda()\n","output_model = Output_Layer(peptide_model, allele_model, peptide_to_output_dim, allele_to_output_dim).cuda()\n","output_model.load_state_dict(torch.load('/content/drive/My Drive/Peptide MHC Project/Trained Models/trained_peptide_allele_sequence_full_model.pt'))\n"],"execution_count":0,"outputs":[{"output_type":"execute_result","data":{"text/plain":["<All keys matched successfully>"]},"metadata":{"tags":[]},"execution_count":66}]},{"cell_type":"code","metadata":{"id":"2csp_Sp-qwKb","colab_type":"code","outputId":"97a42a42-3ed0-488f-c70e-9e1d9a699bee","executionInfo":{"status":"ok","timestamp":1588889059080,"user_tz":420,"elapsed":68587,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":136}},"source":["affinities = []\n","for i in range(corona_peptides.shape[0]):\n","  peptide_affinities = []\n","  for j in range(alleles.shape[0]):\n","    peptide = torch.reshape(corona_embedding_tensor[i, :, :], (1, 15, 100))\n","    allele = torch.reshape(allele_embedding_tensor[j, :, :], (1, 106, 100))\n","    binding_pred = output_model(peptide, allele).item()\n","    peptide_affinities.append(binding_pred)\n","  affinities.append(peptide_affinities)\n","  if i % 100 == 0:\n","    print(i)\n"],"execution_count":0,"outputs":[{"output_type":"stream","text":["0\n","100\n","200\n","300\n","400\n","500\n","600\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"pwTDGJdv13rJ","colab_type":"code","outputId":"cdd6aafb-482e-40cd-dbe3-56e2cf10ceb6","executionInfo":{"status":"ok","timestamp":1588884376023,"user_tz":420,"elapsed":955,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":119}},"source":["n_alleles = 30\n","all_affs = []\n","HLA_A_affs = []\n","HLA_B_affs = []\n","HLA_C_affs = []\n","[all_affs.extend(el) for el in affinities]\n","[HLA_A_affs.extend(el[:10]) for el in affinities]\n","[HLA_B_affs.extend(el[10:20]) for el in affinities]\n","[HLA_C_affs.extend(el[20:]) for el in affinities]\n","\n","avg_aff = sum(all_affs) / len(all_affs)\n","meanHLA_A = sum(HLA_A_affs) / len(HLA_A_affs)\n","meanHLA_B = sum(HLA_B_affs) / len(HLA_B_affs)\n","meanHLA_C = sum(HLA_C_affs) / len(HLA_C_affs)\n","\n","print(\"Mean Binding Affinity between peptides and 30 MHC-I alleles:\", round(avg_aff, 3))\n","print(\"Mean Binding Affinity between peptides and 10 MHC-I alleles with gene:\")\n","print(\"- HLA-A: {}\".format(round(meanHLA_A, 3)))\n","print(\"- HLA-B: {}\".format(round(meanHLA_B, 3)))\n","print(\"- HLA-C: {}\\n\".format(round(meanHLA_C, 3)))\n"],"execution_count":0,"outputs":[{"output_type":"stream","text":["Mean Binding Affinity between peptides and 30 MHC-I alleles: 0.714\n","Mean Binding Affinity between peptides and 10 MHC-I alleles with gene:\n","- HLA-A: 0.549\n","- HLA-B: 0.706\n","- HLA-C: 0.888\n","\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"mXxcZa5gsdYu","colab_type":"code","outputId":"1fa6b143-334a-49aa-8777-fec00dfee33f","executionInfo":{"status":"ok","timestamp":1588885117091,"user_tz":420,"elapsed":590,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":204}},"source":["# Shows binding affinities by percentile\n","x = sorted(all_affs)\n","print(\"25th percentile:\", round(x[int(len(x)*.25)], 3))\n","print(\"50th percentile:\", round(x[int(len(x)*.5)], 3))\n","print(\"75th percentile: %f\\n\" % round(x[int(len(x)*.75)], 3))\n","\n","\n","# Almost all peptides will certainly bind to at least 1 MHC allele\n","n_peptides = len(affinities)\n","near_one_all = []\n","near_zero_all = []\n","either_all = []\n","for pep_affs in affinities:\n","  near_one = []\n","  near_zero = []\n","  either = []\n","  for aff in pep_affs:\n","    near_one.append(aff >= 0.9)\n","    near_zero.append(aff <= 0.1)\n","    either.append(aff >= 0.9 or aff <= 0.1)\n","  near_one_all.append(near_one)\n","  near_zero_all.append(near_zero)\n","  either_all.append(either)\n","\n","pct_near_one = sum([sum(near_one) / n_alleles for near_one in near_one_all]) / n_peptides\n","pct_near_zero = sum([sum(near_zero) / n_alleles for near_zero in near_zero_all]) / n_peptides\n","pct_either = sum([sum(either) / n_alleles for either in either_all]) / n_peptides\n","print(\"Percentage of binding affinities over 0.9: {}%\".format(round(pct_near_one*100, 1)))\n","print(\"Percentage of binding affinities under 0.1: {}%\".format(round(pct_near_zero*100, 1)))\n","print(\"Either (under 0.1 or over 0.9): {}%\".format(round(pct_either*100, 1)))\n","\n","pctHLA_A = sum([sum(in_range[:10]) / 10 for in_range in near_one_all]) / n_peptides\n","pctHLA_B = sum([sum(in_range[10:20]) / 10 for in_range in near_one_all]) / n_peptides\n","pctHLA_C = sum([sum(in_range[20:]) / 10 for in_range in near_one_all]) / n_peptides\n","print(\"Percent over 0.9, by gene:\")\n","print(\"- HLA-A: {}%\".format(round(pctHLA_A*100, 1)))\n","print(\"- HLA-B: {}%\".format(round(pctHLA_B*100, 1)))\n","print(\"- HLA-C: {}%\".format(round(pctHLA_C*100, 1)))\n"],"execution_count":0,"outputs":[{"output_type":"stream","text":["25th percentile: 0.29\n","50th percentile: 0.997\n","75th percentile: 1.000000\n","\n","Percentage of binding affinities over 0.9: 63.4%\n","Percentage of binding affinities under 0.1: 21.0%\n","Either (under 0.1 or over 0.9): 84.4%\n","Percent over 0.9, by gene:\n","- HLA-A: 44.7%\n","- HLA-B: 61.8%\n","- HLA-C: 83.8%\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"1UC3udQjwQ5I","colab_type":"code","outputId":"85f6b35b-4bf3-499e-eccc-01f1a7dacac8","executionInfo":{"status":"ok","timestamp":1588882079738,"user_tz":420,"elapsed":1029,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":238}},"source":["# Finds average affinity by protein\n","# Extracts each protein's amino acid sequence\n","corona = pd.read_csv(\"coronavirus_netchop.csv\")\n","aa_indices = corona.index[corona['pos'] == 1].tolist()\n","n_proteins = len(aa_indices)\n","protein_list = []\n","for i in range(n_proteins):\n","  if i < len(aa_indices) - 1:\n","    end_index = aa_indices[i + 1]\n","  else:\n","    end_index = corona.shape[0]\n","  protein = \"\".join(corona['AA'].iloc[aa_indices[i]:end_index].to_list())\n","  protein_list.append(protein)\n","\n","# Finds which proteins each peptide is in\n","protein_affinities = [[] for _ in range(n_proteins)]\n","for i in range(n_peptides):\n","  peptide = corona_peptides.peptide.iloc[i]\n","  for j in range(n_proteins):\n","    if peptide in protein_list[j]:\n","      protein_affinities[j].append(affinities[i])\n","\n","# Calculates average binding affinity per protein\n","print(\"Number of proteins:\", n_proteins)\n","for i in range(n_proteins):\n","  count = len(protein_affinities[i])\n","  if count == 0: \n","    continue\n","  avg_aff = sum([sum(pep_affs) / len(pep_affs) for pep_affs in protein_affinities[i]]) / count\n","  print(\"Mean Binding Affinity between 30 MHC-I alleles and the {} peptides found in protein {}: {}\"\n","        .format(count, i+1, round(avg_aff, 3)))"],"execution_count":0,"outputs":[{"output_type":"stream","text":["Number of proteins: 14\n","Mean Binding Affinity between 30 MHC-I alleles and the 14 peptides found in protein 2: 0.75\n","Mean Binding Affinity between 30 MHC-I alleles and the 25 peptides found in protein 3: 0.83\n","Mean Binding Affinity between 30 MHC-I alleles and the 4 peptides found in protein 4: 0.728\n","Mean Binding Affinity between 30 MHC-I alleles and the 10 peptides found in protein 5: 0.68\n","Mean Binding Affinity between 30 MHC-I alleles and the 4 peptides found in protein 7: 0.669\n","Mean Binding Affinity between 30 MHC-I alleles and the 1 peptides found in protein 8: 0.451\n","Mean Binding Affinity between 30 MHC-I alleles and the 447 peptides found in protein 9: 0.712\n","Mean Binding Affinity between 30 MHC-I alleles and the 287 peptides found in protein 10: 0.696\n","Mean Binding Affinity between 30 MHC-I alleles and the 98 peptides found in protein 11: 0.711\n","Mean Binding Affinity between 30 MHC-I alleles and the 4 peptides found in protein 12: 0.814\n","Mean Binding Affinity between 30 MHC-I alleles and the 16 peptides found in protein 13: 0.612\n","Mean Binding Affinity between 30 MHC-I alleles and the 2 peptides found in protein 14: 0.721\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"v23RIUjBakie","colab_type":"code","outputId":"3baaa70f-814d-4cf7-bc05-281e3a62c63b","executionInfo":{"status":"ok","timestamp":1588882750536,"user_tz":420,"elapsed":532,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":102}},"source":["# Protein 11 is the Spike Protein\n","count = len(protein_affinities[10])\n","spike_avg = sum([sum(pep_affs) / len(pep_affs) for pep_affs in protein_affinities[10]]) / count\n","print(\"Spike protein (11) mean binding affinity: {}\".format(round(spike_avg, 3)))\n","\n","meanHLA_A = sum([sum(pep_affs[:10]) / 10 for pep_affs in protein_affinities[10]]) / count\n","meanHLA_B = sum([sum(pep_affs[10:20]) / 10 for pep_affs in protein_affinities[10]]) / count\n","meanHLA_C = sum([sum(pep_affs[20:]) / 10 for pep_affs in protein_affinities[10]]) / count\n","print(\"By gene:\")\n","print(\"- HLA-A: {}\".format(round(meanHLA_A, 3)))\n","print(\"- HLA-B: {}\".format(round(meanHLA_B, 3)))\n","print(\"- HLA-C: {}\".format(round(meanHLA_C, 3)))\n"],"execution_count":0,"outputs":[{"output_type":"stream","text":["Spike protein (11) mean binding affinity: 0.711\n","By gene:\n","- HLA-A: 0.518\n","- HLA-B: 0.731\n","- HLA-C: 0.885\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"r9QkDCe811L9","colab_type":"code","outputId":"d311addc-d8de-4bf2-e08f-0842120d6b58","executionInfo":{"status":"ok","timestamp":1588886601430,"user_tz":420,"elapsed":1060,"user":{"displayName":"Jeremy Goldwasser","photoUrl":"","userId":"01547707975030465337"}},"colab":{"base_uri":"https://localhost:8080/","height":425}},"source":["# Breaks peptides apart by length\n","length_affinities = [[] for _ in range(8)]\n","for i in range(n_peptides):\n","  peptide = corona_peptides.peptide.iloc[i]\n","  length_affinities[len(peptide)-8].append(affinities[i])\n","\n","# Calculates mean binding affinity by peptide length\n","for i in range(8):\n","  all = []\n","  [all.extend(peps) for peps in length_affinities[i]]\n","  avg_aff = sum(all) / len(all)\n","  print(\"{} peptides of length {}\".format(len(length_affinities[i]), i+8))\n","  print(\"- Mean binding affinity:\", round(avg_aff, 2))\n","  sorted_affs = sorted(all)\n","  count = len(sorted_affs)\n","  first_idx = [aff >= 0.9 for aff in sorted_affs].index(True)\n","  pct_over = round((count - first_idx) / count * 100)\n","  print(\"- Percentage of binding affinities above 0.9: {}%\".format(pct_over))\n"],"execution_count":0,"outputs":[{"output_type":"stream","text":["100 peptides of length 8\n","- Mean binding affinity: 0.82\n","- Percentage of binding affinities above 0.9: 76%\n","250 peptides of length 9\n","- Mean binding affinity: 0.57\n","- Percentage of binding affinities above 0.9: 45%\n","150 peptides of length 10\n","- Mean binding affinity: 0.7\n","- Percentage of binding affinities above 0.9: 61%\n","25 peptides of length 11\n","- Mean binding affinity: 0.89\n","- Percentage of binding affinities above 0.9: 86%\n","25 peptides of length 12\n","- Mean binding affinity: 0.94\n","- Percentage of binding affinities above 0.9: 90%\n","25 peptides of length 13\n","- Mean binding affinity: 0.98\n","- Percentage of binding affinities above 0.9: 96%\n","25 peptides of length 14\n","- Mean binding affinity: 0.98\n","- Percentage of binding affinities above 0.9: 97%\n","25 peptides of length 15\n","- Mean binding affinity: 0.94\n","- Percentage of binding affinities above 0.9: 93%\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"dT30tHVn3u27","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]},{"cell_type":"code","metadata":{"id":"H2mjFKJSVShW","colab_type":"code","colab":{}},"source":[""],"execution_count":0,"outputs":[]}]}